PL2/PL2-core/src/main/java/nl/tudelft/pl2016gr2/core/algorithms/SplitGraphs.java

```diff
@@ -0,0 +1,151 @@
+package nl.tudelft.pl2016gr2.core.algorithms;
+
+import nl.tudelft.pl2016gr2.model.Node;
+import nl.tudelft.pl2016gr2.model.OriginalGraph;
+import nl.tudelft.pl2016gr2.test.utility.TestId;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+
+/**
+ * Provides functionality for extracting a subgraph of a sequence graph based on a subset of
+ * genomes.
+ * <p>
+ * The resulting subgraph will provide the following characteristics.
+ * <ul>
+ * <li>The nodes in the subgraph are a subset of the nodes in the original graph.</li>
+ * <li>The nodes in the subgraph will have an identical ID as their counterparts in the original
+ * graph.</li>
+ * <li>The nodes will be newly instantiated, meaning that the subgraph datastructure is not backed
+ * by the original graph.</li>
+ * </ul>
+ * </p>
+ *
+ * @author Wouter Smit
+ */
+public class SplitGraphs {
+
+  @TestId(id = "graph_field")
+  private final OriginalGraph mainGraph;
+
+  /**
+   * Instantiates an algorithmic class on a <code>OriginalGraph</code>.
+   * <p>
+   * When {@link #getSubgraph(Collection)} is called, the class will construct a new instance of
+   * the
+   * graph containing only the specified genomes.
+   * </p>
+   *
+   * @param mainGraph The graph to split
+   */
+  public SplitGraphs(OriginalGraph mainGraph) {
+    this.mainGraph = mainGraph;
+  }
+
+
+  /**
+   * Builds a subgraph of the {@link #mainGraph} containing only the nodes that contain (at least)
+   * one of the specified genomes.
+   *
+   * @param genomes The genomes that should be included in the resulting graph
+   * @return An <code>OriginalGraph</code> that is a subgraph of <code>mainGraph</code>
+   * @throws NoSuchElementException Iff the genome collection is not a
+   *                                subset of the main graph's genomes.
+   */
+  public OriginalGraph getSubgraph(Collection<String> genomes) {
+    if (!mainGraph.getGenoms().containsAll(genomes)) {
+      throw new NoSuchElementException("All genomes must be present in the main graph.");
+    }
+    OriginalGraph subGraph = buildSubgraph(genomes);
+    return prune(subGraph, genomes);
+  }
+
+  /**
+   * Creates a new graph containing only the nodes that will be part of the subGraph.
+   * <p>
+   * The nodes are not a deep copy, however the graph object is newly instantiated.
+   * </p>
+   *
+   * @param genomes The genomes that should be included in the subgraph
+   * @return The subgraph
+   */
+  private OriginalGraph buildSubgraph(Collection<String> genomes) {
+    OriginalGraph subGraph = new OriginalGraph();
+    // Add all nodes for every genome
+    for (String genome : genomes) {
+      Iterator<Node> iterator = mainGraph.iterator(genome);
+      // add genome root
+      subGraph.addNode(iterator.next());
+
+      while (iterator.hasNext()) {
+        subGraph.addNode(iterator.next());
```

Comment:
You could create a new node here and add it to subGraph. You can then immediately add the current genome to this node, this is more efficient then doing it afterwards. 
You can then add a check to see if the subgraph already contains a node: if this is the case, get the existing node and add the genome. Else, create a new node and add it to the subgraph.